### PR TITLE
[TreeView] Support item reordering using drag and drop

### LINKED ISSUE #9686: [tree view] Add drag & drop support
An initial issue to keep track of our progress on the drag & drop feature. Note that we won't make it MIT.

## Benchmarks

- https://www.grapecity.com/wijmo/demos/Nav/TreeView/Nodes/DragandDrop/react
- https://ant.design/components/tree#components-tree-demo-draggable
- https://www.treejs.cn/v3/demo.php#_301
- https://www.igniteui.com/tree/drag-and-drop-single-tree
- https://atlaskit.atlassian.com/packages/confluence/tree
- https://wwwendt.de/tech/fancytree/demo/#sample-multi-dnd5.html
- https://ej2.syncfusion.com/documentation/treeview/drag-and-drop
- https://www.youtube.com/watch?v=5SQkOyzZLHM

### PR DESCRIPTION
**Pull request by flaviendelangle**
[Doc preview](https://deploy-preview-12213--material-ui-x.netlify.app/x/react-tree-view/rich-tree-view/ordering/)
Closes #9686

## Requirements for this PR

- #12226
- #12240
- #12610
- #13126
- #13225
- #13804 
- #13953
- #13982

## Basic example

1. Enable the `indentationAtItemLevel` experimental feature (see #13126)
2. Pass the `itemsReordering` prop and the `itemsReordering` experimental feature

```tsx
<RichTreeViewPro
    items={ITEMS}
    itemsReordering
    experimentalFeatures={{ indentationAtItemLevel: true, itemsReordering: true }}
/>
```

## Current status

- [x] Support on `TreeItem`
- [x] Support on `TreeItem2`
- [x] Write doc
- [x] Write tests

### CODE REVIEW

File: docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.js
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.js:1]: Thanks!
What was the reason why you went for the `touchXXX` event instead of `dragXXX`?

[the remark this answers is not in the document]
Comment (cherniavskii) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.js:1]: Did I? I don't recall using touch events

[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.js:1]: `mouseXXX` sorry, which will force you to use `touchXXX` to support mobile from what I understand

[the remark this answers is not in the document]
Comment (cherniavskii) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.js:1]: I don't remember exactly, but it probably didn't work as expected.
I will double check this

[the remark this answers is not in the document]
Comment (cherniavskii) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.js:1]: I used mouse events only for panel resizing. But after reading Matheus' comment in https://github.com/mui/mui-x/issues/9330#issuecomment-2021094965 I switched to pointer events – see https://github.com/cherniavskii/mui-x/blob/pivoting/packages/x-data-grid-premium/src/hooks/utils/useResize.ts

For drag & drop, I used the `draggable` attribute and drag events. I've extracted the `useResize` hook to a separate file to avoid confusion: https://github.com/cherniavskii/mui-x/blob/pivoting/packages/x-data-grid-premium/src/components/GridPivotPanel.tsx

File: packages/x-tree-view/src/useTreeItem2/useTreeItem2.ts [hunk 1]
```diff
@@ -132,6 +167,13 @@ export const useTreeItem2 = <TPlugins extends DefaultTreeViewPlugins = DefaultTr
       onFocus: createRootHandleFocus(externalEventHandlers),
       onBlur: createRootHandleBlur(externalEventHandlers),
       onKeyDown: createRootHandleKeyDown(externalEventHandlers),
+      ...(itemsReordering?.enabled
+        ? {
+            draggable: instance.canItemBeDragged(itemId),
+            onDragStart: createRootHandleDragStart(externalEventHandlers),
+            onDragEnd: createRootHandleDragEnd(externalEventHandlers),
```
Comment (cherniavskii) [inline packages/x-tree-view/src/useTreeItem2/useTreeItem2.ts:174]:
```suggestion
            onDragEnd: createRootHandleDragEnd(externalEventHandlers),
            onDragOver: (event) => event.preventDefault(),
```

To prevent draggable items from returning to their original position:


| Before | After |
| ------- | ----- |
| <video src="https://github.com/mui/mui-x/assets/13808724/73716e27-3d91-4fd7-a6b6-56c8e3688cd7" /> | <video src="https://github.com/mui/mui-x/assets/13808724/80a46f98-eb75-48f6-bdf7-d2c97de15b49" /> |

Comment (flaviendelangle) [reply packages/x-tree-view/src/useTreeItem2/useTreeItem2.ts:174]: That's interesting, I can't reproduce your bug
Are you just dragging and then dropping in another location?

Comment (cherniavskii) [reply packages/x-tree-view/src/useTreeItem2/useTreeItem2.ts:174]: > Are you just dragging and then dropping in another location?

Yes. I can reproduce it in Chromium 123, as well as in Safari 17.4.1 on MacOS 14.4.1

**Comment by benedictchen** [comment]
Cheering for you, brother.  Let's get this in!

頑張れ!

File: packages/x-tree-view/src/TreeItem2DragAndDropOverlay/TreeItem2DragAndDropOverlay.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      },
+    },
+    {
+      props: { action: 'move-to-parent' },
+      style: {
+        marginLeft:
+          'calc(var(--TreeView-indentMultiplier) * calc(var(--TreeView-itemDepth) - 1))' as any,
+        borderBottom: `2px dashed ${(theme.vars || theme).palette.primary.main}`,
+      },
+    },
```
Comment (noraleonte) [inline packages/x-tree-view/src/TreeItem2DragAndDropOverlay/TreeItem2DragAndDropOverlay.tsx:56]: Leaving some suggestions here to improve the styling of the overlay. 👇 

https://github.com/mui/mui-x/assets/72460825/1e27c539-3939-4ddc-8364-6dbbe71ec880


```suggestion
    {
      props: { action: 'make-child' },
      style: {
        marginLeft: 'calc(var(--TreeView-indentMultiplier) * var(--TreeView-itemDepth))',
        borderRadius: theme.shape.borderRadius,
        backgroundColor: alpha((theme.vars || theme).palette.primary.dark, 0.15),
      },
    },
    {
      props: { action: 'reorder-above' },
      style: {
        marginLeft: 'calc(var(--TreeView-indentMultiplier) * var(--TreeView-itemDepth))',
        borderTop: `1px solid ${alpha((theme.vars || theme).palette.grey[900], 0.6)}`,
        ...(theme.palette.mode === 'dark' && {
          borderTop: `1px solid ${alpha((theme.vars || theme).palette.grey[100], 0.6)}`,
        }),
      },
    },
    {
      props: { action: 'reorder-below' },
      style: {
        marginLeft: 'calc(var(--TreeView-indentMultiplier) * var(--TreeView-itemDepth))',
        borderBottom: `1px solid ${alpha((theme.vars || theme).palette.grey[900], 0.6)}`,
        ...(theme.palette.mode === 'dark' && {
          borderBottom: `1px solid ${alpha((theme.vars || theme).palette.grey[100], 0.6)}`,
        }),
      },
    },
    {
      props: { action: 'move-to-parent' },
      style: {
        marginLeft:
          'calc(var(--TreeView-indentMultiplier) * calc(var(--TreeView-itemDepth) - 1))' as any,
        borderBottom: `1px solid ${alpha((theme.vars || theme).palette.grey[900], 0.6)}`,
        ...(theme.palette.mode === 'dark' && {
          borderBottom: `1px solid ${alpha((theme.vars || theme).palette.grey[900], 0.6)}`,
        }),
      },
    },
```

Comment (flaviendelangle) [reply packages/x-tree-view/src/TreeItem2DragAndDropOverlay/TreeItem2DragAndDropOverlay.tsx]: I pushed your CSS :+1:

File: packages/x-tree-view/src/models/items.ts [hunk 3]
```diff
@@ -4,3 +4,9 @@ export type TreeViewItemId = string;
 export type TreeViewBaseItem<R extends {} = { id: string; label: string }> = R & {
   children?: TreeViewBaseItem<R>[];
 };
+
+export type TreeViewItemsReorderingAction =
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/models/items.ts:8]: This type need to be on the community package because it is used in `TreeItem2DragAndDropOverlay` which is on the community package (to avoid creating a different Tree Item component on the pro package)

File: packages/x-tree-view/src/TreeItem2DragAndDropOverlay/TreeItem2DragAndDropOverlay.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        borderBottom: `1px solid ${alpha((theme.vars || theme).palette.grey[900], 0.6)}`,
+        ...(theme.palette.mode === 'dark' && {
+          borderBottom: `1px solid ${alpha((theme.vars || theme).palette.grey[900], 0.6)}`,
+        }),
+      },
+    },
+  ],
+}));
+
+function TreeItem2DragAndDropOverlay(props: TreeItem2DragAndDropOverlayProps) {
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/TreeItem2DragAndDropOverlay/TreeItem2DragAndDropOverlay.tsx:63]: This is on the community package because it allows to have a single Tree Item component shared across all plans.
For now I did not try to add any shiny abstraction to have it in the pro package, I think having a dummy component here is the best trade-off

If we want to go a little further, we can see how to inject the style from the pro plan since it's almost all the this component currently have, but I don't think it's worth the effort for now

File: packages/x-tree-view/src/internals/useTreeView/useTreeView.ts [hunk 5]
```diff
@@ -122,6 +124,9 @@ export const useTreeView = <Plugins extends readonly TreeViewPlugin<TreeViewAnyP
   contextValue.runItemPlugins = (itemPluginProps) => {
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/internals/useTreeView/useTreeView.ts:124]: I think I'll move all the `runItemPlugin` calls in a standalone hook to be able to document it more thoroughly, but it would make the review harder so it stays here for now.

The idea of these new enhancers is to have each plugin responsible for generating its props passed to the various slots.
It allows to have `TreeItem2DragAndDropOverlay` being super dumb.
If we scale this to more plugins, we might have to have a smarter merging process for callbacks and refs, but I don't think it's a top priority.

File: docs/data/tree-view/rich-tree-view/ordering/OnlyReorderFromDragHandle.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    if (!onDragStart) {
+      return;
+    }
+
+    onDragStart(event);
+    event.dataTransfer.setDragImage(event.target.parentElement, 0, 0);
+  };
+
+  return (
+    <TreeItem2Provider itemId={itemId}>
```
Comment (flaviendelangle) [inline docs/data/tree-view/rich-tree-view/ordering/OnlyReorderFromDragHandle.js:66]: I don't think we can skip re-creating the whole Tree Item here in order to get the content props and re-dispatch them to the drag handle.

One way of doing it would be to add `handleDragStart` and `handleDragStop` on `useTreeItem2Utils`.
That way, this demo could just override its `content` slot to render the default `TreeItem2Content` with the `children` it receives plus the added icon (and pass the handlers to it from `useTreeItem2Utils`).
It could then use `slotProps` and `defaultMuiPrevented` to remove the behaviors from the root.

Basically it would be closer to [this demo](https://mui.com/x/react-tree-view/rich-tree-view/customization/#limit-expansion-to-icon-container), which would be super nice because if in the future we add editing, or anything else, it would just work out of the box.

But adding it to `useTreeItem2Utils` is not trivial because it means this hook must be able to be enriched from the plugins somehow, and right now it does not.
That will be a nice follow up.

File: packages/x-tree-view-pro/src/internals/plugins/useTreeViewItemsReordering/useTreeViewItemsReordering.types.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+export interface UseTreeItem2ContentSlotPropsFromItemsReordering {
+  onDragEnter?: MuiCancellableEventHandler<React.DragEvent>;
+  onDragOver?: MuiCancellableEventHandler<React.DragEvent>;
+}
+
+export interface UseTreeItem2DragAndDropOverlaySlotPropsFromItemsReordering
+  extends TreeItem2DragAndDropOverlayProps {}
+
+declare module '@mui/x-tree-view/useTreeItem2' {
```
Comment (flaviendelangle) [inline packages/x-tree-view-pro/src/internals/plugins/useTreeViewItemsReordering/useTreeViewItemsReordering.types.ts:149]: One limitation of using `declare` here is that if you use one `RichTreeView` and one `RichTreeViewPro`, and if this `RichTreeView` use a custom Tree Item, then this custom Tree Item will have a drag & drop props on the value returned by `getRootProps` and the other functions.

I think it's fine

File: docs/data/tree-view/rich-tree-view/ordering/OnlyReorderFromDragHandle.js (hunk 6 above)
Comment (noraleonte) [reply docs/data/tree-view/rich-tree-view/ordering/OnlyReorderFromDragHandle.js:76]: Makes a lot of sense to me. I think the DX is good for now and we can think about a follow-up for the utils

File: docs/data/tree-view/rich-tree-view/ordering/ordering.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+---
+
+# Rich Tree View - Ordering [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
+
+<p class="description">Drag and drop your items to reorder them.</p>
+
+:::success
+To be able to reorder items, you first have to enable the `indentationAtItemLevel` experimental feature.
+When this flag is enabled, the indentation of nested items is applied by the item itself instead of its ancestors.
+This allows correctly placing the drag & drop overlay.
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/ordering.md:17]: I totally removed those sentences in favor of a link to the dedicated doc section (that did not existed when I opened this PR :angel: )

File: docs/data/tree-view/rich-tree-view/ordering/ordering.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+If you are building your custom Tree Item, make sure that you do not apply any custom `padding-left` on your Tree Item Content component.
+If you want to change the indentation value, you can use the `itemChildrenIndentation` prop on the Tree View component.
+:::
+
+## Enable drag & drop re-ordering
+
+You can enable the drag & drop re-ordering of items by setting the `itemsReordering` prop to `true`:
+
+{{"demo": "DragAndDrop.js"}}
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/ordering.md:31]: What do you mean by "expansion breaks"
Would you expect the targeted item to be automatically expended when we drop the item?

File: docs/data/tree-view/rich-tree-view/ordering/ordering.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+## Limit the re-ordering
+
+By default, all the items are reorderable.
+You can prevent the re-ordering of some items using the `isItemReorderable` prop.
+The following example shows how to only allow re-ordering of the leaves using the [`getItemOrderedChildrenIds`](/x/react-tree-view/rich-tree-view/items/#get-an-items-children-by-id) API method.
+
+{{"demo": "OnlyReorderLeaves.js"}}
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/ordering.md:39]: The example of reordering leaves is not great because it is here to showcase `isItemReorderable` so it limits which items you can re-order, not where you can re-order them :/
Which means right now you can even drop `@mui/x-data-grid` inside `@mui/x-date-pickers`...


Is the little red cursor in the video something that is really visible in your browser?
Because since you are able to drop in any item, I'm surprised this is visible...

[the remark this answers is not in the document]
Comment (noraleonte) [reply docs/data/tree-view/rich-tree-view/ordering/ordering.md:39]: > Is the little red cursor in the video something that is really visible in your browser?

Yes, in both Chrome and Firefox, that's why I said that it's confusing.

File: docs/data/tree-view/rich-tree-view/ordering/ordering.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/ordering.md:31]: Oh!
Very nice catch!
I'm not updating the `TreeViewItemMeta['expandable']` attribute of neither the old parent nor the new parent, this should be a fairly easy fix to do

[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/ordering.md:31]: Fixed

**Review by noraleonte — APPROVED** [review]
Congrats on this huge effort! 🎉 It's looking great! 🔥

File: docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.tsx [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    children: [{ id: 'tree-view-community', label: '@mui/x-tree-view' }],
+  },
+];
+
+export default function DragAndDrop() {
+  return (
+    <Box sx={{ minHeight: 352, minWidth: 300 }}>
+      <RichTreeViewPro
+        items={ITEMS}
+        itemsReordering
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.tsx:41]: You participated in the discussion on Slack :see_no_evil: https://mui-org.slack.com/archives/C040YT7KB6H/p1721220863107809

[the remark this answers is not in the document]
Comment (LukasTy) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.tsx:41]: Fair point... I somehow managed to miss this fact. 😆 
In my head, I had an understanding that we'd have:
```tsx
<RIchTreeViewPro
  experimentalFeatures={{
    itemReordering: true,
  }}
/>
```
which would be changed to after it becomes stable:
```diff
<RIchTreeViewPro
-  experimentalFeatures={{
-    itemReordering: true,
-  }}
+  itemReordering 
/>
```

[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.tsx:41]: That was not an option is proposed indeed

The main advantage of the current version is that people will have no BC when migrating to v8.
But this is a super subjective decision for sure

**Comment by flaviendelangle** [comment]
I opened a PR for the env check because it's trivial to fix and doing it before the release is better

For the rest we can do a follow up

> I believe that clicking Esc should cancel the dragging and releasing the mouse afterwards should not finish it.

Fully agree, I'll fix that

> It would be nice to avoid the animation of the dragging item returning (floating) to the original place before finishing the dragging action.

Could you record your screen with this behavior visible to be sure I'm understand what you are referring to?

File: docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (LukasTy) [reply docs/data/tree-view/rich-tree-view/ordering/DragAndDrop.tsx:41]: Well, yeah, that is the benefit of duplicating the props. 👍 
Let's see how the community receives it. 👌

**Comment by LukasTy** [comment]
> > I believe that clicking Esc should cancel the dragging and releasing the mouse afterwards should not finish it.
> 
> Fully agree, I'll fix that

Makes sense. 👌 


> Could you record your screen with this behavior visible to be sure I'm understand what you are referring to?


https://github.com/user-attachments/assets/0d75835c-9a13-4dcd-8a9f-a56b30e83e7c

I'm circling over the somewhat strange behavior—resembling a canceled drag. 🙈
